app/interactors/albums.js

```
@@ -0,0 +1,26 @@
+const errors = require('../errors');
+const albums = require('../services/albums');
+const db = require('../models');
+const errorMessages = require('../constants/errorsMessages');
+const logger = require('../logger');
+
+exports.userBuysAlbum = (albumId, userId) => {
+  let albumToBuy = null;
+  return albums
+    .getAlbum(albumId)
+    .then(album => {
+      logger.info(`Retrieved album from external service to buy ${JSON.stringify(album)}`);
+      albumToBuy = album;
+      return db.albumsByUser.findAndCountAll({ where: { userId, albumId } }).catch(err => {
+        throw errors.databaseError(err.message);
+      });
+    })
+    .then(albumUser => {
+      if (albumUser.count >= 1) {
+        throw errors.usarAlreadyHasAlbum(errorMessages.userAlreadyHasAlbum);
+      }
+      return db.albumsByUser.create({ userId, albumId, albumTitle: albumToBuy.title }).catch(err => {
+        throw errors.databaseError(err.message);
+      });
+    });
+};
```

Ok, a few things here:

1) You can validate if album-user relation is already created before fetching the album. Therefore you can throw an error faster and avoid an unnecessary external api call.

2) In both then statements you finish you are returning same db error. Keep in mind:
```javascript
return asyncTask()
  .then(() => {
    return anotherAsyncTask().catch(err => {
      throwDatabaseError();
    })
  })
  .then(() => {
    return yetAnotherAsyncTask().catch(err => {
      throwDatabaseError();
    })
  });

// It's the same to do:
return asyncTask()
  .then(() => {
    return anotherAsyncTask();
  })
  .then(() => {
    return yetAnotherAsyncTask();
  })
  .catch(err => {
    throwDatabaseError();
  });
```

---

I think the number 2 comment will overwrite the error `errors.usarAlreadyHasAlbum`.

---

Yes, tried that and the error was overriten. Will change the order but not the error handling